modules/internal/ChapelArray.chpl

```diff
@@ -3010,11 +3010,48 @@ module ChapelArray {
     return b;
   }
 
+  //
+  // These control an optimization in which copying an array from one
+  // locale to another will also copy the array's domain if we believe
+  // it's safe to do so.  This optimization is currently off by
+  // default because it was completed close to Chapel 2.1 and we
+  // wanted more time to live with it before having it on by default.
+  //
+  config param localizeConstDomains = true,
+               debugLocalizedConstDomains = false;
+
   pragma "init copy fn"
   proc chpl__initCopy(const ref rhs: [], definedConst: bool) {
-    pragma "no copy"
-    var lhs = chpl__coerceCopy(rhs.type, rhs, definedConst);
-    return lhs;
+    //
+    // create a local copy of a remote array's domain if...
+    // - the optimization is enabled
+    // - we're running using multiple locales
+    // - the array copy we're declaring is 'const' (implying that
+    //   its domain can't be re-assigned while it's alive) or the
+    //   original domain is 'const' (implying that it can never
+    //   be re-assigned)
+    // - the domain is on a remote locale
+    //
+    const localize = (localizeConstDomains &&
+                      numLocales > 1 &&
+                      (definedConst || rhs.domain.definedConst) &&
+                      rhs.domain._value.locale != here);
```

Comment:
Does it make sense to add `compiledForSingleLocale()` here as well? This is a param and could avoid a branch when COMM=none. Arguably not important for now and you could probably just log it as future work in https://github.com/chapel-lang/chapel/issues/25238

Reply:
That's a good idea.  We could even potentially have both param and non-param aspects of this `const` split apart and to have the conditional be something like:

```chapel
if paramLocalize && localize {
  …
} else {
  …
}
```

s.t. if `paramLocalize` was false, we'd fold the conditional?  I like the idea of forking this off to its own PR, primarily to keep working on things that aren't done yet and adding it to the follow-on issue if you're good with that (though of all the things on that issue, I may do it first—possibly even tomorrow if time permits).